lib/charms/opensearch/v0/opensearch_nodes_exclusions.py

```diff
@@ -70,55 +72,134 @@ def delete_current(self) -> None:
                 self._scope, ALLOCS_TO_DELETE, ",".join(current_allocations)
             )
 
+    def _removed_units_to_cleanup(self) -> Optional[List[str]]:
+        """Deletes all units that have left the cluster via Juju.
+
+        This method ensures we keep a small list of voting exclusions at all times.
+        """
+        deployment_desc = self._charm.opensearch_peer_cm.deployment_desc()
+        if not deployment_desc or deployment_desc.typ not in [
+            DeploymentType.MAIN_ORCHESTRATOR,
+            DeploymentType.FAILOVER_ORCHESTRATOR,
+        ]:
+            return []
+
+        short_id = self._charm.opensearch_peer_cm.deployment_desc().app.short_id
+        # We do not need to add our own unit as this code is running on it!
+        peers = self._charm.model.get_relation(PeerRelationName)
+        peers = [] if not peers else peers.units
+
+        cms = set(
+            [
+                unit.name + "." + short_id
+                for relation in self._charm.model.relations.get(
+                    PeerClusterOrchestratorRelationName
+                )
+                for unit in relation.units
+            ]
+            + [unit.name + "." + short_id for unit in peers]
+        )
+
+        cleanup_nodes = []
+        for node in self._fetch_voting_exclusions():
+            if node not in cms:
+                cleanup_nodes.append(node)
+        return cleanup_nodes
+
     def cleanup(self) -> None:
         """Delete all exclusions that failed to be deleted."""
-        need_voting_cleanup = self._charm.peers_data.get(self._scope, VOTING_TO_DELETE, False)
-        if need_voting_cleanup and self._delete_voting():
+        need_voting_cleanup = set(
+            self._charm.peers_data.get(self._scope, VOTING_TO_DELETE, "").split(",")
+            + self._removed_units_to_cleanup()
+        )
+        if (
+            need_voting_cleanup
+            and self._delete_voting(to_remove=need_voting_cleanup)
+            and not self._fetch_voting_exclusions()  # called after the entire deletion, must be empty now
```

Comment:
not sure we should be too harsh here, in case the user sets some manually - we should perhaps track only the ones we set from our charm

Reply:
Fair enough. The only concern I can see is that we have 10x votes by default... If we consider the user may try using it, then we may end up not having enough votes!

We can raise that as an exception, ofc.